### PR TITLE
Add device discovery method

### PR DESCRIPTION
**Pull request by piotrbulinski**
This PR adds a method to detect Flexit devices in the local network. It only works on Linux, as it uses `/proc/net/route` to find out the local network broadcast addresses.

This feature was requested in #13

### CODE REVIEW

File: flexit_bacnet/bacnet.py [hunk 1]
```diff
@@ -504,3 +524,124 @@ async def write(self, device_property: DeviceProperty, value: Any):
...
+
+# Flexit uses a proprietary service defined by Siemens to discover devices on the local network.
+VENDOR_ID_SIEMENS = 7
+SERVICE_NUMBER_DISCOVERY = 515
+SERVICE_NUMBER_IDENTIFICATION = 516
+
+# As I cannot find the specification for the service parameters,
+# will use what I've captured them from the Flexit app using Wireshark.
+# It seems to contain a UUID and some other "random" value,
+# but will use a fixed value for now and worry about it if anyone ever complains.
```
Comment (lellky) [inline flexit_bacnet/bacnet.py:537]: 😄

**Comment by lellky** [comment]
I can't get it to work. I'm running Ubuntu 20.04.5 LTS with Python 3.10.9 (and 3.8.10).

Trying with `PYTHONPATH=. python3 examples/discover.py` gives

```
Exception in callback _SelectorTransport._call_connection_lost(None)
handle: <Handle _SelectorTransport._call_connection_lost(None)>
Traceback (most recent call last):
  File "/usr/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.10/asyncio/selector_events.py", line 742, in _call_connection_lost
    self._protocol.connection_lost(exc)
  File "/home/lellky/projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 466, in connection_lost
    self.done.set_result(True)
asyncio.exceptions.InvalidStateError: invalid state
Exception in callback _SelectorTransport._call_connection_lost(None)
```

**Comment by piotrbulinski** [comment]
@lellky thanks for testing, I've refactored the discovery mechanism and I would appreciate if you could give it another try.

**Comment by lellky** [comment]
@piotrbulinski now the code seems to work, but it can't find my device. They are on the same local network.

```
lellky@duke:~/projects/flexit_bacnet$ PYTHONPATH=. python3 examples/discover.py
No devices found
```

**Comment by lellky** [comment]
Screenshot 2023-12-20 at 10 58 55

File: flexit_bacnet/bacnet.py [hunk 2]
```diff
@@ -602,46 +601,45 @@ def _is_discovery_response(response: bytes) -> bool:
...
-            if device_ip in devices:
-                continue
-
-            devices.append(bacnet_request.src[0])
-
-    return devices
+    response_ips = set()
+
+    # Create a UDP socket
+    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
```
Comment (lellky) [inline flexit_bacnet/bacnet.py:633]: Maybe also add `socket.IPPROTO_UDP` as parameter?

**Comment by piotrbulinski** [comment]
@lellky thanks, it worked for me on workstation connected vie ethernet, but it was failing on WiFi connection. I've made some changes, to account for network latency/dropped packets. Please try again and let's hope it works this time :)

**Comment by lellky** [comment]
The code looks good. But I still get "No devices found". From what I can see in Wireshark the communication seems to work, but flexit_bacnet can for some reason not hear the response (?).

192.168.2.18 is my Ubuntu where flexit_bacnet runs. 192.168.2.14 is my Flexit Nordic. The device_info example still works.

Screenshot 2023-12-20 at 20 03 45

**Comment by lellky** [comment]
Okay, I don't know why it does not work on my Ubuntu machine. But now it works on my Macbook Pro M1 ✅ 

```
lellky@Jonass-MacBook-Air flexit_bacnet % PYTHONPATH=. python3 examples/discover.py
Found devices:
192.168.2.14
```

Awesome! Great work!

**Review by lellky — APPROVED** [review]
LGTM

**Review by lellky — APPROVED** [review]

**Comment by piotrbulinski** [comment]
> Okay, I don't know why it does not work on my Ubuntu machine. But now it works on my Macbook Air M1 ✅

Do you have any firewall configured on this machine by any chance?

**Comment by lellky** [comment]
> > Okay, I don't know why it does not work on my Ubuntu machine. But now it works on my Macbook Air M1 ✅
> 
> @lellky Do you have any firewall configured on this machine by any chance?

Actually, maybe I have. I will investigate and if I have one I will try to turn it off and see if that works.

**Pull request merged**